### PR TITLE
Ensemble with Dask

### PR DESCRIPTION
**Pull request by franchuterivera**
Moving ensemble construction to dask

### CODE REVIEW

File: autosklearn/ensemble_builder.py [hunk 1]
```diff
@@ -30,26 +31,206 @@
...
+
+        # Add the result of the run
+        # On the next while iteration, no references to
+        # ensemble builder object, so it should be garbage collected to
+        # save memory while waiting for resources
+        # Also, notice how ensemble nbest is returned, so we don't waste
+        # iterations testing if the deterministic predictions size can
+        # be fitted in memory
+        try:
+            result = EnsembleBuilder(
```
Comment (mfeurer) [inline autosklearn/ensemble_builder.py:172]: If I understand the code correctly, this now has to read all predictions from scratch every time in order to compute the scores?

File: autosklearn/ensembles/ensemble_selection.py [hunk 2]
```diff
@@ -27,6 +27,17 @@ def __init__(
...
         self.random_state = random_state
 
+    def __getstate__(self) -> Dict[str, Any]:
+        # Cannot serialize a metric if
+        # it is user defined.
+        # That is, if doing pickle dump
+        # the metric won't be the same as the
+        # one in __main__. we don't use the metric
+        # in the EnsembleSelection so this should
+        # be fine
```
Comment (mfeurer) [inline autosklearn/ensembles/ensemble_selection.py:37]: I disagree on this - we use `self.metric` to compute the scores. Does this issue happen in the examples on custom metrics?

File: test/test_automl/test_automl.py [hunk 3]
```diff
@@ -168,7 +197,7 @@ def test_delete_non_candidate_models(self):
         # Assert at least one model file has been deleted and that there were no
         # deletion errors
         log_file_path = glob.glob(os.path.join(
-            backend_api.temporary_directory, 'AutoML(' + str(seed) + '):*.log'))
+            backend_api.temporary_directory, 'autosklearn.ensemble_builder.log'))
```
Comment (mfeurer) [inline test/test_automl/test_automl.py:200]: Would it be possible to have multiple processes write to the same logfile or would that require large amounts of work?

File: test/test_automl/test_automl.py [hunk 4]
```diff
@@ -43,8 +51,24 @@ def setUp(self):
...
         self.automl._delete_output_directories = lambda: 0
 
+    @classmethod
+    def setUpClass(self):
+        dask.config.set({'distributed.worker.daemon': False})
+        self.client = dask.distributed.Client(
+            dask.distributed.LocalCluster(
+                n_workers=1,
+                processes=False,
+                threads_per_worker=2,
```
Comment (mfeurer) [inline test/test_automl/test_automl.py:64]: Does this mean all tests are now executed with two workers in parallel?

Comment (franchuterivera) [reply test/test_automl/test_automl.py]: I have created a fixture so that a global client is created once, and we supply task to this unique client. 

With multiple clients, pytest seem to struggle with:
-> Logging: I think some stdout/stderror capturing is redirected from pytest, and with multiple clients there are multiple threadlocking jobs that at the end of the pytest run are simply not closed. No matter if I call client.shutdown, client.close.


It is cleaner to have a single client which we can track for debug if needed.

File: test/test_automl/test_automl.py (hunk 3 above)
Comment (franchuterivera) [reply test/test_automl/test_automl.py]: Done!

File: autosklearn/ensemble_builder.py (hunk 1 above)
Comment (franchuterivera) [reply autosklearn/ensemble_builder.py:186]: I changed this with a pickled memory as per discussed offline. Please let me know if I misunderstood the problem statement.

File: autosklearn/automl.py [hunk 5]
```diff
@@ -619,11 +670,22 @@ def fit(
...
+            # Notice we provided a kill signal and also the maximum allowed time, so
+            # it is likely that the TimeoutError won't happen. In case it does, we notify
+            # the user
+            try:
+                dask.distributed.wait(proc_ensemble, timeout=self.ensemble_timeout_patience)
+                self.ensemble_performance_history = proc_ensemble.result()
+            except dask.distributed.TimeoutError:
+                self._logger.error("The ensemble process unexpectedly hanged..."
+                                   "Please check the log file for errors.")
+                self._dask_client.cancel(proc_ensemble)
```
[the remark this answers is not in the document]
Comment (franchuterivera) [reply autosklearn/automl.py:692]: No, in the unfortunate case this cancel crashes. The motivation for this is not to  leave a job always running in a client, which will virtually leave the worker unusable

File: autosklearn/automl.py [hunk 6]
```diff
@@ -499,35 +538,47 @@ def fit(
...
+                self._memory_limit,  # pynisher memory limit
+                self._seed,  # random state
             )
-            self._proc_ensemble.start()
+
+            # We expect the ensemble process to be time driven, and honor the maximum
+            # time provided via time_left_for_ensemble. As a safety measure, we also
+            # provide an event signal that stops the process before the next iteration
+            # of ensemble builder process
+            self.ensemble_event_killer = dask.distributed.Event(self.ensemble_event_killer)
```
[the remark this answers is not in the document]
Comment (franchuterivera) [reply autosklearn/automl.py]: Done!

File: autosklearn/automl.py [hunk 7]
```diff
@@ -221,6 +240,27 @@ def __init__(self,
         # After assigning and checking variables...
         # self._backend = Backend(self._output_dir, self._tmp_dir)
 
+    def _create_dask_client(self):
+        self._is_dask_client_internally_created = True
+        processes = False
```
[the remark this answers is not in the document]
Comment (franchuterivera) [reply autosklearn/automl.py:246]: The only problem is if you requested 1 job. If you create a client with 1 job and process true, in my laptop, the client creation goes crazy and forces me to terminate python immediately. But also, I think this will allow us to run on a no multiple cores for debug purposes.

This can be changed to force jobs to be always more than 2 and process equal true always. Let me know your preferences.

[the remark this answers is not in the document]
Comment (mfeurer) [reply autosklearn/automl.py:246]: I like the changes, they make sense.

File: autosklearn/ensemble_builder.py [hunk 8]
```diff
@@ -210,39 +427,37 @@ def __init__(
...
                 break
-            self.time_left = time_left
             safe_ensemble_script = pynisher.enforce_limits(
-                wall_time_in_s=int(time_left),
+                wall_time_in_s=int(time_left + time_buffer),
                 mem_in_mb=self.memory_limit,
                 logger=self.logger
             )(self.main)
-            safe_ensemble_script()
+            safe_ensemble_script(time_left, iteration)
```
Comment (mfeurer) [inline autosklearn/ensemble_builder.py:460]: I think the safeguards below which half the memory usage will no longer work together with the new code structure. As the ensemble builder is re-constructed in every iteration, it can't remember the value for nbest and this needs to be done above in the for loop.

Comment (franchuterivera) [reply autosklearn/ensemble_builder.py:460]: Hi Matthias, the nbest is remembered through this line:

https://github.com/automl/auto-sklearn/blob/8fc9d3090887d0bb5383000f1e1cb72757f1a38d/autosklearn/ensemble_builder.py#L213

File: autosklearn/ensemble_builder.py [hunk 9]
```diff
@@ -384,6 +385,33 @@ def __init__(
...
+        self.ensemble_memory_file = os.path.join(
+            self.backend.temporary_directory,
+            'ensemble_read_preds.pkl'
+        )
+        if os.path.exists(self.ensemble_memory_file):
+            try:
+                with (open(self.ensemble_memory_file, "rb")) as memory:
+                    self.read_preds = pickle.load(memory)
+            except Exception as e:
+                self.logger.critical(
```
[the remark this answers is not in the document]
Comment (franchuterivera) [reply autosklearn/ensemble_builder.py:407]: Will do this in next commit!

File: autosklearn/ensemble_builder.py [hunk 10]
```diff
@@ -384,6 +385,33 @@ def __init__(
...
+        # scores for self.read_preds
+        # is too computationally expensive
+        # As the ensemble builder is stateless
+        # (every time the ensemble builder gets resources
+        # from dask, it builds this object from scratch)
+        # we save the state of this dictionary to memory
+        # and read it if available
+        self.ensemble_memory_file = os.path.join(
+            self.backend.temporary_directory,
+            'ensemble_read_preds.pkl'
```
[the remark this answers is not in the document]
Comment (franchuterivera) [reply autosklearn/ensemble_builder.py:400]: Will do this in next commit!

File: test/conftest.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
+import pytest
```
[the remark this answers is not in the document]
Comment (franchuterivera) [reply test/conftest.py:1]: Will do this in next commit!

File: test/test_ensemble_builder/test_ensemble.py [hunk 12]
```diff
@@ -607,18 +686,19 @@ def test_get_identifiers_from_run_history_log_loss(self, exists):
...
         self.backend = BackendMock()
-        self.backend.temporary_directory = '/tmp/ensemble_test'
-        os.makedirs(self.backend.temporary_directory, exist_ok=True)
 
     def tearDown(self):
-        shutil.rmtree(self.backend.temporary_directory)
+        logfiles = glob.glob(os.path.join(
+            self.backend.temporary_directory, '*.log'))
+        for logfile in logfiles:
+            os.unlink(logfile)
```
[the remark this answers is not in the document]
Comment (franchuterivera) [reply test/test_ensemble_builder/test_ensemble.py:694]: I will add information explaining why this is ended, which is because of the logging we create.

File: test/test_ensemble_builder/test_ensemble.py [hunk 13]
```diff
@@ -452,6 +463,74 @@ def testLimit(self):
...
+    def test_read_pickle_read_preds(self):
+        """
+        This procedure test that we save the read predictions before
+        destroying the ensemble builder and that we are able to read
+        them safely after
+        """
+        ensbuilder = EnsembleBuilder(
+            backend=self.backend,
+            dataset_name="TEST",
+            task_type=3,  # Multilabel Classification
```
[the remark this answers is not in the document]
Comment (franchuterivera) [reply test/test_ensemble_builder/test_ensemble.py:499]: Will do this in next commit!

File: autosklearn/ensembles/ensemble_selection.py (hunk 2 above)
Comment (franchuterivera) [reply autosklearn/ensembles/ensemble_selection.py:37]: We discussed offline:

- We can proceed with removing the metrics from the dump pickle object. Examples work and unit test is not failing. 
- We will update the ensemble selection fit method to print a better message in case metric is None.

File: test/test_ensemble_builder/test_ensemble.py [hunk 14]
```diff
@@ -607,18 +686,19 @@ def test_get_identifiers_from_run_history_log_loss(self, exists):
 class EnsembleProcessBuilderTest(unittest.TestCase):
     def setUp(self):
         self.backend = BackendMock()
-        self.backend.temporary_directory = '/tmp/ensemble_test'
-        os.makedirs(self.backend.temporary_directory, exist_ok=True)
 
     def tearDown(self):
-        shutil.rmtree(self.backend.temporary_directory)
+        logfiles = glob.glob(os.path.join(
```
[the remark this answers is not in the document]
Comment (franchuterivera) [reply test/test_ensemble_builder/test_ensemble.py:691]: I added a comment for this in the next push. Please let me know if still does not make sense.

[the remark this answers is not in the document]
Comment (mfeurer) [reply test/test_ensemble_builder/test_ensemble.py:698]: Sorry, I cannot follow. Could you please write this in still more detail?

File: autosklearn/ensemble_builder.py (hunk 8 above)
Comment (mfeurer) [reply autosklearn/ensemble_builder.py:460]: I see. Do we have a test for this? Also, do we delete temporary file that might be too large now?

Also, what's the else in the return statement returning exactly and when does it happen?

File: examples/60_search/example_successive_halving.py [hunk 15]
```diff
@@ -118,10 +119,11 @@ def get_smac_object(
 
 automl = autosklearn.classification.AutoSklearnClassifier(
     time_left_for_this_task=40,
-    per_run_time_limit=5,
-    tmp_folder='/tmp/autosklearn_sh_example_tmp',
-    output_folder='/tmp/autosklearn_sh_example_out',
+    per_run_time_limit=10,
+    tmp_folder='/tmp/autosklearn_sh_example_tmp_cv',
```
Comment (mfeurer) [inline examples/60_search/example_successive_halving.py:123]: I find those names slightly hard to follow. Could you please use `01`, `02` etc as trailing ways to differentiate.

**Comment by mfeurer** [comment]
Thanks a lot for making the ensemble process working starting new tasks which now show up in the task stream (see below). Also, the ensemble builder now nicely runs on more than on the next free worker. 

However, there's an unfortunate caveat, the ensemble builder can right now actually run twice at the same time. Would it be possible to prevent this?
image

File: test/test_automl/test_estimators.py [hunk 16]
```diff
@@ -26,650 +28,599 @@
...
+    log_file_path = glob.glob(os.path.join(
+        tmp_dir, 'AutoML*.log'))[0]
+    predictions = automl.predict(X_test)
+    assert predictions.shape == (50, 3), extract_msg_from_log(log_file_path)
+    assert count_succeses(automl.cv_results_) > 0,  extract_msg_from_log(log_file_path)
+
+    score = f1_macro(Y_test, predictions)
+    assert score >= 0.9, extract_msg_from_log(log_file_path)
 
-        pickle.dumps(automl_fitted)
```
Comment (mfeurer) [inline test/test_automl/test_estimators.py:671]: Hey, I think something went wrong during the merge or rebase, and the test that this can be pickled is now missing.

File: test/test_ensemble_builder/test_ensemble.py [hunk 17]
```diff
@@ -27,821 +28,706 @@
...
-        ])
-        # Now the weights have 2 zero weights and 2 non-zero weights,
-        # which is incompatible.
-        ensemble.weights_ = [0.6, 0.0, 0.0, 0.4]
-
-        with self.assertRaises(ValueError):
-            ensemble.predict(per_model_pred)
-
-
-class SingleBestTest(unittest.TestCase):
```
Comment (mfeurer) [inline test/test_ensemble_builder/test_ensemble.py:513]: Are these tests still in the new version?

File: autosklearn/ensemble_builder.py [hunk 18]
```diff
@@ -202,28 +209,27 @@ def ensemble_builder_process(
...
+                exception_traceback = traceback.format_exc()
+                error_message = repr(e)
+                logger.critical(exception_traceback)
+                logger.critical(error_message)
 
         # If pynisher kills a run, the result
         # might be None -- so no new timestamp info
-        done_futures = [f for f in futures if f.done()]
-        for future in done_futures:
+        if future.done():
```
Comment (mfeurer) [inline autosklearn/ensemble_builder.py:232]: This looks like it could fail - assume that we're in the 1st iteration and fail in the try/except above. Now, future is still `None`.

File: test/test_automl/base.py [hunk 19]
```diff
@@ -9,6 +9,34 @@
...
+    return os.linesep.join(content)
+
+
+def count_succeses(cv_results):
+    return np.sum(
+        [status in ['Success', 'Success (but do not advance to higher budget)']
+         for status in cv_results['status']]
+    )
+
+
```
Comment (mfeurer) [inline test/test_automl/base.py:39]: The class below can probably deleted, too?

Comment (franchuterivera) [reply test/test_automl/base.py]: Thanks you are right. I will fix that in the next push

File: autosklearn/ensemble_builder.py (hunk 18 above)
Comment (franchuterivera) [reply autosklearn/ensemble_builder.py]: Thanks, I have fixed that.

File: test/test_ensemble_builder/test_ensemble.py (hunk 17 above)
Comment (franchuterivera) [reply test/test_ensemble_builder/test_ensemble.py:513]: Hi Matthias, yes. In detail:
```
old_test_name->new_test_name
test_get_identifiers_from_run_history_log_loss->test_get_identifiers_from_run_history
test_get_identifiers_from_run_history_accuracy->test_get_identifiers_from_run_history
```

I am using the new fixtures, so that is why the name changed

File: test/test_automl/test_estimators.py (hunk 16 above)
Comment (franchuterivera) [reply test/test_automl/test_estimators.py:671]: thanks for catching it, I have fixed this.

**Comment by franchuterivera** [comment]
A new push was added to make the ensemble process be a thread from the main loop rather than a dask future, which might not be allocated to the task pool (That is, experimental results show that the ensemble process resides on the same worker it was originally sent).

Here I am using a thread instead of multiprocessing process as one cannot send 'easily' a dask job from a multiprocessing fork, which can be reproduced in:

```
import dask
import dask.distributed

import multiprocessing
import time


def second_client(address):
    print(f"Here in second client")
    client2 = dask.distributed.Client(address=address)
    future = client2.submit(lambda: 5 * 5)
    print(future.result())
    print(f"Done in second client")


if __name__ == '__main__':
    client = dask.distributed.Client(n_workers=2, processes=1, threads_per_worker=1)
    proc = multiprocessing.Process(
        target=second_client, args=(client.scheduler_info()['address'], ))
    future = client.submit(lambda: 5 * 5)
    print(future.result())
    print(f"Then, what doesn't work:")
    proc.start()

```

File: autosklearn/ensemble_builder.py [hunk 20]
```diff
@@ -250,9 +248,7 @@ def ensemble_builder_process(
             )
 
         # Don't take resources while waiting
-        dask.distributed.secede()
         time.sleep(sleep_duration)
```
Comment (mfeurer) [inline autosklearn/ensemble_builder.py:251]: I think this part could be slightly reworked. Previously, if the ensemble script constructed a new ensemble, it wouldn't sleep. It would only sleep if there were no new predictions available. So I guess if you could add an additional flag to the fit_and_return_ensemble that indicates whether an ensemble was constructed, that would bring back the original behavior.

File: autosklearn/ensemble_builder.py [hunk 21]
```diff
@@ -229,7 +227,7 @@ def ensemble_builder_process(
 
         # If pynisher kills a run, the result
         # might be None -- so no new timestamp info
-        if future.done():
+        if future is not None and future.done():
```
Comment (mfeurer) [inline autosklearn/ensemble_builder.py:230]: Could we actually wait for the future to be done? Maybe with a timeout of 1s to also check for the ensemble termination request? I believe this would be closer to how dask actually wants things to be done.

**Comment by mfeurer** [comment]
Superseded by #992
